osrf/gazebo
# Re-implement GpuLaser with cube map to lift vFOV limit

# Motivation

The current implementation of the GpuRaySensor limits the vertical FOV to 90 degree. This makes it impossible to simulate some types of Lidar sensors with it (e.g. dome Lidar).

This PR re-implements the GpuRaySensor with a different internal method to lift this limitation. The vertical FOV can now be up to 180 degree, so that a full sphere can be covered.

# How to test this change

It should now be possible to use the GpuRaySensor with vertical FOVs of up to 180 degrees. You can use this provided example world file for testing with a full sphere (hFOV=360°, vFOV=180°) lidar sensor: 
[full_sphere_gpu_lidar.world.txt](https://github.com/enwaytech/gazebo-old/files/8388881/full_sphere_gpu_lidar.world.txt) (remove .txt extension)

Apart from that also test other common configurations, especially edge cases (e.g. only 1 vertical ray).

# Internal method

Previously an internal camera was rotated around the spin axis of the sensor to capture depth images in a first pass, while in a second pass those were warped to match the laser ray geometry. Due to the limited vertical FOV of this camera, the overall vertical FOV was limited.

The new method uses the internal camera to capture depth images in up to six orientations to build a depth cube map. Those up to six depth images are then used for the laser readings. Compared to the previous method a few things were simplified: The camera aspect ratio is always 1 (square depth images), the camera orientations are fixed and only one single rendering pass is required (previously two passes).
Each face of this virtual cube corresponds to a depth image taken by the internal camera. For reference, these figures show the cube map setup. _Azimuth_ corresponds to the horizontal, _elevation_ to the vertical angle of a laser ray:

Side view            |  Top view
:-------------------------:|:-------------------------:
![Cube Map (side view)](https://user-images.githubusercontent.com/10380243/159897711-fb1501ff-a21f-42aa-83ca-bc9f4338e83f.jpg) | ![Cube Map (top view)](https://user-images.githubusercontent.com/10380243/159897714-4e176bef-293d-4638-8cd6-cccddc90d7d8.jpg)

Only the required faces of the cube are rendered to avoid unnecessary computation.

# Limitations

As before this method has the same resolution issues in the case of a very shallow angle of incidence, where you may see "steps" in the scan. This issue is inherent in using a rasterized depth image as a base for the range measurements. To mitigate this issue the resolution of the internally used depth camera can be increased, however that comes with a longer processing time. I introduced a minimum texture size of 1024x1024 to avoid a strong effect when the number of samples is low, I found this value to be a good compromise. This minimal texture size was already there before and set to 2048 but there it was using fewer cameras.

Compare these two images of the VLP-16 sensor (`roslaunch velodyne_description example.launch gpu:=true`) to see that the results are "similarly good/bad" (note: the angle of incidence on the ground plane is extremely shallow here, so the effect is clearly visible. In the average case it should look much better):

Current implementation            |  New cube map implementation
:-------------------------:|:-------------------------:
![VLP-16_old_version](https://user-images.githubusercontent.com/10380243/161043796-dd8e90c5-613b-4948-9922-350591fbcb00.png) |  ![VLP-16_new_version](https://user-images.githubusercontent.com/10380243/161043791-41a0f81c-49e3-4007-94ba-e9a26ef15639.png)

# Code

The changes mainly take place in the `GpuRaySensor` and `GpuLaser` classes. I left the public interface of these classes unchanged, unsure if that is sufficient to meet the API compatibility criterion. I'm also not really sure about how to test the ABI compatibility. Here I could use some help by the reviewers.

A few unit tests were added, the existing unit and integration tests seem to pass.
